StorageUtility/Windows/Device.cpp

```diff
@@ -0,0 +1,97 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+</License>
+*/
+#include "Device.h"
+#include "StorageUtility.h"
+
+namespace vtStor
+{
+
+cDevice::cDevice(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData)
+{
+    AllocateMemories(DevInfoData, DevInterfaceData, DevDetailData, SizeOfDevDetailData);
+    CopyMemories(DevInfoData, DevInterfaceData, DevDetailData, SizeOfDevDetailData);
+}
+
+cDevice::~cDevice()
+{
+    DeallocateMemories();
+}
+
+void cDevice::Data(std::unordered_map<eDeviceDataType, void*>& Data)
+{
+    Data.insert({ eDeviceDataType::PSP_DEVINFO_DATA, m_DevInfoData });
+    Data.insert({ eDeviceDataType::PSP_DEVICE_INTERFACE_DATA, m_DevInterfaceData });
+    Data.insert({ eDeviceDataType::PSP_INTERFACE_DEVICE_DETAIL_DATA, m_DevDetailData });
+}
+
+DeviceHandle cDevice::Handle()
+{
+    eErrorCode errorCode;
+    DeviceHandle deviceHandle;
+
+    errorCode = GetStorageDeviceHandle(m_DevDetailData->DevicePath, deviceHandle);
+    if (eErrorCode::None != errorCode)
+    {
+        throw std::exception("Failed to create device handle", GetLastError());
+    }
+    return deviceHandle;
+}
+
+eErrorCode cDevice::GetStorageDeviceHandle(const String& DevicePath, HANDLE& Handle)
+{
+    Handle =
+        CreateFile(
+        DevicePath.c_str(),
+        GENERIC_READ | GENERIC_WRITE,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        NULL,
+        OPEN_EXISTING,
+        FILE_ATTRIBUTE_NORMAL,
+        NULL
+        );
+
+    if (INVALID_HANDLE_VALUE == Handle)
+    {
+        return(eErrorCode::Io);
+    }
+
+    return(eErrorCode::None);
+}
+
+void cDevice::AllocateMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData)
+{
+    m_DevInfoData = (PSP_DEVINFO_DATA)HeapAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, DevInfoData->cbSize);
+    m_DevInterfaceData = (PSP_DEVICE_INTERFACE_DATA)HeapAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, DevInterfaceData->cbSize);
+    m_DevDetailData = (PSP_INTERFACE_DEVICE_DETAIL_DATA)HeapAlloc(GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, SizeOfDevDetailData);
+}
+
+void cDevice::CopyMemories(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData, U32 SizeOfDevDetailData)
+{
+    memcpy(m_DevInfoData, DevInfoData, DevInfoData->cbSize);
+    memcpy(m_DevInterfaceData, DevInterfaceData, DevInterfaceData->cbSize);
+    memcpy(m_DevDetailData, DevDetailData, SizeOfDevDetailData);
```

Comment:
1. Please use the secure version of these functions: `memcpy_s` and  `wmemcpy_s`.
2. Decide whether or not to use the wide character version in code.